File: backend/app/routers/tasks.py
"""任务：列表 / 登记 / 修改 / 详情 / 运行（模拟一次结果并入库）。"""
import random
from datetime import datetime, timezone

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy import delete, func, select
from sqlalchemy.ext.asyncio import AsyncSession

from ..db import get_session
from ..models import Task, TaskRun, User
from ..permissions import require
from ..schedule import compute_next_run, validate_cron
from ..schemas import TaskCreate, TaskUpdate
from ..serializers import task_to_dict

router = APIRouter(prefix="/api", tags=["tasks"])

TASK_TYPES = {
    "review": "评价采集",
    "chat": "聊天采集",
    "draw": "绘图采集",
    "conv": "转化率监控",
    "monitor": "数据监控",
    "report": "报表任务",
}

# 只暴露引用名，密钥本体由服务端环境/密管持有
SECRET_REFS = [
    {"ref": "ops-callback-key", "description": "回调网关签名密钥"},
    {"ref": "tmall-open-api", "description": "天猫开放平台凭证"},
    {"ref": "jd-open-api", "description": "京东开放平台凭证"},
]

STATUSES = ["运行中", "待流转", "已阻塞", "本轮已完成"]


def _validate(body: TaskCreate | TaskUpdate) -> None:
    if body.schedule_kind == "cron":
        if not body.cron_expr or not validate_cron(body.cron_expr):
            raise HTTPException(status_code=400, detail="cron 表达式无效")
    elif body.schedule_kind == "interval":
        if not body.interval_minutes:
            raise HTTPException(status_code=400, detail="请填写间隔分钟数")
    else:
        raise HTTPException(status_code=400, detail="调度方式只能是 cron 或 interval")
    if body.type not in TASK_TYPES:
        raise HTTPException(status_code=400, detail="任务类型无效")
    if body.callback_url and not body.callback_url.startswith(("http://", "https://")):
        raise HTTPException(status_code=400, detail="回调地址必须是 http(s) URL")
    if body.callback_secret_ref and body.callback_secret_ref not in {s["ref"] for s in SECRET_REFS}:
        raise HTTPException(status_code=400, detail="secretRef 不存在")
    if body.points_done > body.points_total:
        raise HTTPException(status_code=400, detail="已完成点位不能超过总点位")
    if isinstance(body, TaskUpdate) and body.status is not None and body.status not in STATUSES:
        raise HTTPException(status_code=400, detail="状态无效")


@router.get("/meta", dependencies=[Depends(require("task:read"))])
async def meta():
    return {"task_types": TASK_TYPES, "statuses": STATUSES, "secret_refs": SECRET_REFS}


@router.get("/tasks", dependencies=[Depends(require("task:read"))])
async def list_tasks(session: AsyncSession = Depends(get_session)):
    tasks = (
        (await session.execute(select(Task).order_by(Task.next_run_at.is_(None), Task.next_run_at)))
        .scalars()
        .all()
    )
    return [task_to_dict(t) for t in tasks]


@router.get("/tasks/{task_id}", dependencies=[Depends(require("task:read"))])
async def get_task(task_id: int, session: AsyncSession = Depends(get_session)):
    task = await session.get(Task, task_id)
    if task is None:
        raise HTTPException(status_code=404, detail="任务不存在")
    return task_to_dict(task, with_runs=True)


async def _next_code(session: AsyncSession) -> str:
    n = (await session.execute(select(func.count(Task.id)))).scalar_one()
    return f"FL-{3100 + n}"


@router.post("/tasks", status_code=201)
async def create_task(
    body: TaskCreate,
    session: AsyncSession = Depends(get_session),
    _: User = Depends(require("task:create")),
):
    _validate(body)
    task = Task(**body.model_dump(), code=await _next_code(session))
    task.next_run_at = compute_next_run(task.schedule_kind, task.cron_expr, task.interval_minutes)
    session.add(task)
    await session.commit()
    await session.refresh(task)
    return task_to_dict(task, with_runs=True)


@router.put("/tasks/{task_id}")
async def update_task(
    task_id: int,
    body: TaskUpdate,
    session: AsyncSession = Depends(get_session),
    _: User = Depends(require("task:update")),
):
    task = await session.get(Task, task_id)
    if task is None:
        raise HTTPException(status_code=404, detail="任务不存在")
    _validate(body)
    for key, value in body.model_dump(exclude={"status"}).items():
        setattr(task, key, value)
    if body.status is not None:
        task.status = body.status
    task.next_run_at = compute_next_run(task.schedule_kind, task.cron_expr, task.interval_minutes)
    await session.commit()
    await session.refresh(task)
    return task_to_dict(task, with_runs=True)


def _mock_result(task: Task) -> tuple[bool, str]:
    if random.random() < 0.2:
        fails = [
            f"失败 · 接口超时（{random.randint(2, max(2, task.points_total))} 个点位无数据）",
            "失败 · 回调网关返回 502",
            "失败 · 触发反爬限流，等待冷却窗口",
        ]
        return False, random.choice(fails)
    n = random.randint(300, 2000)
    texts = {
        "review": f"成功 · 入库 {n:,} 条评价",
        "chat": f"成功 · 采集 {n:,} 条会话",
        "draw": f"成功 · 回传 {random.randint(1, task.points_total)} 店设计稿",
        "conv": f"成功 · 全站转化率 {random.uniform(2.8, 4.2):.1f}%",
        "monitor": f"成功 · 巡查 {task.points_total} 个点位，无告警",
        "report": f"成功 · 已出 {random.randint(1, task.points_total)} 份报表",
    }
    return True, texts.get(task.type, f"成功 · 处理 {n:,} 条记录")


@router.post("/tasks/{task_id}/run")
async def run_task(
    task_id: int,
    session: AsyncSession = Depends(get_session),
    _: User = Depends(require("task:run")),
):
    task = await session.get(Task, task_id)
    if task is None:
        raise HTTPException(status_code=404, detail="任务不存在")

    now = datetime.now(timezone.utc)
    ok, text = _mock_result(task)
    session.add(
        TaskRun(
            task_id=task.id,
            ran_at=now,
            ok=ok,
            result_text=text,
            duration_ms=random.randint(5_000, 90_000),
        )
    )

    if ok:
        if task.points_done < task.points_total:
            task.points_done = min(task.points_total, task.points_done + random.randint(1, 3))
        task.status = "本轮已完成" if task.points_done >= task.points_total else "运行中"
    else:
        task.status = "已阻塞"
    task.next_run_at = compute_next_run(
        task.schedule_kind, task.cron_expr, task.interval_minutes, base=now
    )

    # 备份保留最近 keep_runs 次运行
    await session.flush()
    keep_ids = (
        (
            await session.execute(
                select(TaskRun.id)
                .where(TaskRun.task_id == task.id)
                .order_by(TaskRun.ran_at.desc())
                .limit(task.keep_runs)
            )
        )
        .scalars()
        .all()
    )
    await session.execute(
        delete(TaskRun).where(TaskRun.task_id == task.id, TaskRun.id.not_in(keep_ids))
    )

    await session.commit()
    await session.refresh(task)
    return task_to_dict(task, with_runs=True)

File: apps/tms/src/lib/tasks.ts
/** 任务领域逻辑：校验、序列化、模拟运行、备份裁剪。 */
import { asCount, execute, query, queryOne } from "./db";
import { SECRET_REFS, STATUSES, TASK_TYPES } from "./meta";
import { computeNextRun, isValidCron } from "./schedule";
import { ApiError } from "./rbac";

export type TaskInput = {
  title: string;
  description?: string;
  type?: string;
  owner_name?: string;
  channel?: string;
  points_done?: number;
  points_total?: number;
  points_note?: string;
  schedule_kind?: string;
  cron_expr?: string | null;
  interval_minutes?: number | null;
  callback_url?: string | null;
  callback_timeout_ms?: number;
  callback_retries?: number;
  callback_secret_ref?: string | null;
  keep_runs?: number;
  status?: string;
};

export type TaskRow = Record<string, unknown> & { id: number };

function num(v: unknown, fallback: number): number {
  const n = Number(v);
  return Number.isFinite(n) ? Math.round(n) : fallback;
}

export function normalizeInput(body: TaskInput): TaskInput {
  const t: TaskInput = {
    title: String(body.title ?? "").trim(),
    description: String(body.description ?? ""),
    type: String(body.type ?? "review"),
    owner_name: String(body.owner_name ?? ""),
    channel: String(body.channel ?? ""),
    points_done: num(body.points_done, 0),
    points_total: num(body.points_total, 1),
    points_note: String(body.points_note ?? ""),
    schedule_kind: String(body.schedule_kind ?? "cron"),
    cron_expr: body.cron_expr ? String(body.cron_expr).trim() : null,
    interval_minutes: body.interval_minutes ? num(body.interval_minutes, 0) : null,
    callback_url: body.callback_url ? String(body.callback_url).trim() : null,
    callback_timeout_ms: num(body.callback_timeout_ms, 5000),
    callback_retries: num(body.callback_retries, 3),
    callback_secret_ref: body.callback_secret_ref ? String(body.callback_secret_ref) : null,
    keep_runs: num(body.keep_runs, 10),
    status: body.status ? String(body.status) : undefined,
  };

  if (!t.title) throw new ApiError(400, "请填写任务名称");
  if (!(t.type! in TASK_TYPES)) throw new ApiError(400, "任务类型无效");
  if (t.schedule_kind === "cron") {
    if (!t.cron_expr || !isValidCron(t.cron_expr)) throw new ApiError(400, "cron 表达式无效");
  } else if (t.schedule_kind === "interval") {
    if (!t.interval_minutes || t.interval_minutes < 1) throw new ApiError(400, "请填写间隔分钟数");
  } else {
    throw new ApiError(400, "调度方式只能是 cron 或 interval");
  }
  if (t.callback_url && !/^https?:\/\//.test(t.callback_url)) {
    throw new ApiError(400, "回调地址必须是 http(s) URL");
  }
  if (t.callback_secret_ref && !SECRET_REFS.some((s) => s.ref === t.callback_secret_ref)) {
    throw new ApiError(400, "secretRef 不存在");
  }
  if (t.callback_timeout_ms! < 100 || t.callback_timeout_ms! > 120000) {
    throw new ApiError(400, "回调超时需在 100–120000 毫秒之间");
  }
  if (t.callback_retries! < 0 || t.callback_retries! > 10) throw new ApiError(400, "重试次数需在 0–10 之间");
  if (t.keep_runs! < 1 || t.keep_runs! > 100) throw new ApiError(400, "备份份数需在 1–100 之间");
  if (t.points_total! < 1) throw new ApiError(400, "总点位至少为 1");
  if (t.points_done! < 0 || t.points_done! > t.points_total!) {
    throw new ApiError(400, "已完成点位需在 0 与总点位之间");
  }
  if (t.status && !STATUSES.includes(t.status as (typeof STATUSES)[number])) {
    throw new ApiError(400, "状态无效");
  }
  return t;
}

export async function serializeTask(row: TaskRow, withRuns = false) {
  const runs = await query<Record<string, unknown>>(
    "SELECT id, ran_at, ok, result_text, duration_ms FROM task_runs WHERE task_id = ? ORDER BY ran_at DESC LIMIT ?",
    [row.id, withRuns ? 100 : 1],
  );
  const toRun = (r: Record<string, unknown>) => ({ ...r, ok: Boolean(r.ok) });
  return {
    ...row,
    last_run: runs.length ? toRun(runs[0]) : null,
    ...(withRuns ? { runs: runs.map(toRun) } : {}),
  };
}

export async function getTask(id: number): Promise<TaskRow> {
  const row = await queryOne<TaskRow>("SELECT * FROM tasks WHERE id = ?", [id]);
  if (!row) throw new ApiError(404, "任务不存在");
  return row;
}

export async function nextTaskCode(): Promise<string> {
  const row = await queryOne<{ n: number }>("SELECT COUNT(*) AS n FROM tasks");
  return `FL-${3100 + asCount(row?.n)}`;
}

function randInt(min: number, max: number): number {
  return Math.floor(Math.random() * (max - min + 1)) + min;
}

export function mockRunResult(task: TaskRow): { ok: boolean; text: string } {
  const total = task.points_total as number;
  if (Math.random() < 0.2) {
    const fails = [
      `失败 · 接口超时（${randInt(2, Math.max(2, total))} 个点位无数据）`,
      "失败 · 回调网关返回 502",
      "失败 · 触发反爬限流，等待冷却窗口",
    ];
    return { ok: false, text: fails[randInt(0, fails.length - 1)] };
  }
  const n = randInt(300, 2000).toLocaleString("en-US");
  const texts: Record<string, string> = {
    review: `成功 · 入库 ${n} 条评价`,
    chat: `成功 · 采集 ${n} 条会话`,
    draw: `成功 · 回传 ${randInt(1, total)} 店设计稿`,
    conv: `成功 · 全站转化率 ${(2.8 + Math.random() * 1.4).toFixed(1)}%`,
    monitor: `成功 · 巡查 ${total} 个点位，无告警`,
    report: `成功 · 已出 ${randInt(1, total)} 份报表`,
  };
  return { ok: true, text: texts[task.type as string] ?? `成功 · 处理 ${n} 条记录` };
}

/** 运行一次（模拟结果入库），并把备份裁剪到最近 keep_runs 次。 */
export async function runTask(id: number): Promise<TaskRow> {
  const task = await getTask(id);
  const now = new Date();
  const { ok, text } = mockRunResult(task);

  await execute(
    "INSERT INTO task_runs (task_id, ran_at, ok, result_text, duration_ms) VALUES (?, ?, ?, ?, ?)",
    [id, now.toISOString(), ok ? 1 : 0, text, randInt(5000, 90000)],
  );

  let pointsDone = task.points_done as number;
  const total = task.points_total as number;
  let status: string;
  if (ok) {
    if (pointsDone < total) pointsDone = Math.min(total, pointsDone + randInt(1, 3));
    status = pointsDone >= total ? "本轮已完成" : "运行中";
  } else {
    status = "已阻塞";
  }
  const nextRun = computeNextRun(
    task.schedule_kind as string,
    task.cron_expr as string | null,
    task.interval_minutes as number | null,
    now,
  );
  await execute("UPDATE tasks SET points_done = ?, status = ?, next_run_at = ?, updated_at = ? WHERE id = ?", [
    pointsDone,
    status,
    nextRun,
    now.toISOString(),
    id,
  ]);

  await execute(
    `DELETE FROM task_runs WHERE task_id = ? AND id NOT IN (
       SELECT id FROM (
         SELECT id FROM task_runs WHERE task_id = ? ORDER BY ran_at DESC LIMIT ?
       ) kept
     )`,
    [id, id, task.keep_runs as number],
  );

  return getTask(id);
}
